### PR TITLE
feat(evm): use completely separated storage sections in multifork

### PR DESCRIPTION
**Pull request by mattsse**
## Motivation
with #1715 we added multifork support, with some assumptions/limitations:
* only one _local_ memory area, which meant that _write operations_ were performed in the same memory area regardless of forking

this has turned out to be an undesired and potential pitfall.

This rewrites how storage is tracked:
* each fork is completely separate, _write operations_ are performed only in the fork's own separate memory
* the storage of the account of the `sender` and the `contract` remains persistent across swaps: e.g: changes to the test contract's own variables are persistent when selecting another fork

## Solution
* store multiple `forkId -> ForkDB` pairs
* also keep track and update `Subroutine` when selecting forks
* drastically simplify the `FuzzWrapper` which is now a `Cow<Backend>`

### CODE REVIEW

**Review by gakonst — APPROVED** [review]
clean, anything else we should test beyond storage values? 

cc @hexonaut, @aureliusbtc & @mds1 for any user testing we can get before merging in case we're introducing any edge cases here.

**Comment by hexonaut** [comment]
I'm getting a non-descript error in this test at this line with this PR (previously working): https://github.com/makerdao/xdomain/blob/add-optimism/packages/dss-bridge/src/tests/domains/optimism/OptimismIntegration.t.sol#L292

**Comment by mattsse** [comment]
@gakonst I think this is good to merge, but maybe we should wait until I updated all the docs, also need another round of revm debugging to figure out how to enable diagnostics for calls on non-existent errors.

But from a functionality point of view it is complete and can be merged

**Comment by gakonst** [comment]
Ack - reviewing today. Want to give some thinking on whether we can do this without the persist cheats

**Comment by mattsse** [comment]
@gakonst figured out the problem with diagnostics, 

now we return meaningful revert reasons if someone calls a contract that only exists on another fork:

https://github.com/foundry-rs/foundry/blob/df7ecd865c9cea3e52b6f2fb833092602f985320/forge/src/multi_runner.rs#L1210-L1213

File: testdata/cheats/Fork2.t.sol
Comment (gakonst) [inline testdata/cheats/Fork2.t.sol:116]: should this be commented out?

**Review by gakonst — APPROVED** [review]

File: testdata/cheats/Fork2.t.sol
Comment (mattsse) [reply testdata/cheats/Fork2.t.sol]: deleted since no longer used in this test, reverts with a meaningful message are now tested separately

**Comment by mattsse** [comment]
I'm going to revise all multifork docs next. Should we hold off on merging until tonight's night is built so we can ship it with tomorrow's nightly?

**Comment by gakonst** [comment]
Yep - let's give this one final pass and merge/ship tomorrow.

**Comment by gakonst** [comment]
@hexonaut @mds1 merged, see how to use [here](https://github.com/foundry-rs/book/pull/488/files). This should allow us to have contracts that persist across forks beyond the default test contract and msg.sender.

**Pull request merged**
